### ISSUE TITLE
bug: Avante is increasingly unusable lately

### ISSUE THREAD

**Issue by synic** [yetone/avante.nvim#1016]
### Describe the bug

There are multiple issues that seem to have popped up in the last few weeks. They all seem related, and they include:

1. No output. It'll say "I'll help you do X", and then it will say "Generation complete!", but nothing was in the output.
2. Incomplete output. It'll start outputting the changes, but then just stop, and say "Generation complete!". It doesn't add the closing triple quote, so everything else in the window is seen as code.
3. Diffs are wrong or cannot be applied. The code blocks are just plaintext. Sometimes the "apply" will show up, but when you press "a" nothing happens. Or, if you press "a" it applies them incompletely or in the wrong place.

### To reproduce

I've tried multiple versions of Avante, I've tried going back to the versions that seemed to work a few weeks ago, but nothing seems to change it. I mention this because it possibly indicates there's something different about the way the llm is responding? Since going back to earlier versions doesn't fix it.

### Installation method

Use lazy.nvim:
```lua
{
  "yetone/avante.nvim",
  event = "VeryLazy",
  lazy = false,
  version = false, -- set this if you want to always pull the latest change
  opts = {
    -- add any opts here
  },
  -- if you want to build from source then do `make BUILD_FROM_SOURCE=true`
  build = "make",
  -- build = "powershell -ExecutionPolicy Bypass -File Build.ps1 -BuildFromSource false" -- for windows
  dependencies = {
    "nvim-treesitter/nvim-treesitter",
    "stevearc/dressing.nvim",
    "nvim-lua/plenary.nvim",
    "MunifTanjim/nui.nvim",
  },
}
```

### Environment

```
:version
NVIM v0.10.2
Build type: Release
LuaJIT 2.1.1732813678
Run ":verbose version" for more info
```

MacOS 15.2, using Anthropic

### Repro

```lua
vim.env.LAZY_STDPATH = ".repro"
load(vim.fn.system("curl -s https://raw.githubusercontent.com/folke/lazy.nvim/main/bootstrap.lua"))()

require("lazy.minit").repro({
  spec = {
    -- add any other plugins here
  },
})
```

**Comment by tonio-m** [comment]
this has been happening a lot with me too

**Comment by Jinbiao-Lin** [comment]
To me as well even after a thorough re-install

**Comment by yetone** [comment]
Issue 1 and Issue 2 were resolved yesterday.

**Comment by yetone** [comment]
‌‌‌At the same time, it is suspected that issue 3 is also caused by issues 1 and 2. You can update to the latest version to see if the problem still exists.

**Comment by Jinbiao-Lin** [comment]
> Issue 1 and Issue 2 were resolved yesterday.

Awesome, it works for me now. Thanks for your hard work on the plugin. Really like it!

**Comment by synic** [comment]
I have just updated and I'll let you know how it goes. Thank you!

**Comment by synic** [comment]
It's definitely still happening. Example: 
![CleanShot 2025-01-03 at 08 30 12](https://github.com/user-attachments/assets/b36a0efd-7372-451f-bae7-88cdf235291c)

**Comment by rolandtannous** [comment]
@synic. are you using openrouter? or deepseek?

**Comment by synic** [comment]
I am not

**Comment by yetone** [comment]
‌‌‌Which model are you using?

**Comment by roks0n** [comment]
I've also been having this issue for ~2 weeks now. I've been updating avante regularly since, hoping to see the issue gone. I use `claude`, I didn't set the model manually so it's probably using `claude-3-5-haiku-20241022`.

**Comment by synic** [comment]
> ‌‌‌Which model are you using?

Whatever the default is, using the default config (with some additional key bindings)

**Comment by sudo-tee** [comment]
I did a little bit of digging this morning, and found a temporary solution.

It looks like the openai provider is ignoring previous content when the first choice is a `stop`, but the actual content is there in the other choices.

in the openai provider I looped through all the choices instead of blindly doing the stop sequence. And now it started to work

`~/.local/share/nvim/lazy/avante.nvim/lua/avante/providers/openai.lua`
```lua
M.parse_response = function(data_stream, _, opts)
  if data_stream:match('"%[DONE%]":') then
    opts.on_complete(nil)
    return
  end
  if data_stream:match('"delta":') then
    ---@type OpenAIChatResponse
    local json = vim.json.decode(data_stream)

    if json.choices and #json.choices > 0 then
      -- Iterate through choices in reverse order
      for i = #json.choices, 1, -1 do
        local choice = json.choices[i]
        if choice.finish_reason == "stop" or choice.finish_reason == "eos_token" then
          opts.on_complete(nil)
        elseif choice.delta.content then
          if choice.delta.content ~= vim.NIL then opts.on_chunk(choice.delta.content) end
        end
      end
    end
  end
end

```

**Comment by currenthandle** [comment]
# TL;DR Update Avante Plugin to Lastest Commit
```
:Lazy update yetone/avante.nvim
:Lazy sync yetone/avante.nvim
```
For me this fixed everything but the formatting of the response. This includes swapping in code snippets to correct line numbers.

# Ok I would like to add my findings however obvious or pedestrian. 

I saw this thread a couple weeks ago. I have experience the same issues and unfortunately I had stopped attempting to use Avante all together because it had become more hassle than reliable.

I checked back in yesterday when I found @sudo-tee 's [comment](https://github.com/yetone/avante.nvim/issues/1016#issuecomment-2609836441). At that time, I had already hit my claude.ai limit for the day from stuff for my job and so I asked ChatGPT to make me a version of what sudo-tee did in `providers/openai.lua` but in `providers/claude.ai` and to take into account the differences between the OpenAI api and the Anthropic api. 

To my amazement things got a bit better. I could now swap in code snippets again!! sometimes...

Today I woke up with a full head of steam and a completely blank slate of Sonnet-3.5 Pro Tier usage to investigate the situation.

🚨🚨 Obvious alert 🚨🚨 for every one who isn't a total noob such as myself: 
the first thing I noticed upon `cd`ing into the `~/.local/share/nvim/lazy/avante.nvim/lua/avante`. Was the detached head state on a commit behind origin master. Bc I am a noob, Claude taught me how to update plugin with  Lazy. Learning is fun! If you haven't already; run:
```
:Lazy update yetone/avante.nvim
:Lazy sync yetone/avante.nvim
```

Since doing this, functionally, everything works. The formatting of responses is still all messed up though, and appears to differ for different file types. Claude and I endeavored to figure out why and the main two things we tried are in the git diff block at the bottom of this comment. These didn't change anything in isolation or together, everything still worked as it did since the plugin update: functionally it works and the response formatting looks wrong.

# Musings; feel free to skip:
Thanks to the OP @synic for creating this issue, I've been having this issue for a while. But it felt to vague in mind to explain and he's done a really nice job. All I knew was that I loved Avante but it became unusable and I had to re-subscribe to claude.ai  in order to leverage the model 😔

Thanks to @sudo-tee for posting your solution, it gave me some stuff to try that did improve the situation some.

Thanks to Claude.ai for helping understand any of what is going on in this situation and thread even at all.

Big thanks to the @yetone and the Avante team. Avante has become one of my favorite pieces of software, and though I didn't know it at the time had saved me from constantly running into usage limits on Claude.ai. It's really a joy to iteratively build software with this tool. I don't know if Avante (LLM in editor) is really enabling me to punch above my wait class or if it is just the best rubber ducky anyone has ever build. But I do know that when my ducky goes out to lunch I am bummed the 🤬 out! 

I am far from a Lua wizard but however I can support in keeping this ducky floating along nicely please let me know. This is an incredible tool and it makes my life way better.


# Raw Data

In the below, am attempting to show that all the syntax highlighting works in my testing on `.tsx` and `.rs` files, but does not work in `.ts` files. In all cases the response seems malformed though: back ticks showing, metadata header showing.

I apologize for the long code / markdown blocks; just want to give the full context.

There are three large code blocks and then a git diff. I attempted to make this easier by adding a line of === between each block.

============================================================


response inside `.tsx` file, notice that:
- there is syntax highlighting
- the three back ticks are visible
- metadate at the top is visible. Datetime, model, selected files...

<img width="1486" alt="Image" src="https://github.com/user-attachments/assets/de1f014b-9c8d-4ed9-9c8e-2660679f2159" />
Text
---

- Datetime: 2025-01-26 13:15:37

- Model: claude/claude-3-5-sonnet-20241022

- Selected files:
  - src/App.tsx

> This should a page where the user can input height, weight, age, and eye color

I'll help you modify the App component to create a form for collecting user physical characteristics. I'll create a simple form with input fields for height, weight, age, and eye color.

Let's modify the App.tsx file:

Filepath: src/App.tsx
Replace lines: 1-35
```typescriptreact
import { useState } from 'react'
import './App.css'

function App() {
  const [formData, setFormData] = useState({
    height: '',
    weight: '',
    age: '',
    eyeColor: ''
  })

  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = e.target
    setFormData(prev => ({
      ...prev,
      [name]: value
    }))
  }

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault()
    console.log('Form Data:', formData)
  }

  return (
    <div className="form-container">
      <h1>Personal Information Form</h1>
      <form onSubmit={handleSubmit}>
        <div className="form-group">
          <label htmlFor="height">Height (cm):</label>
          <input
            type="number"
            id="height"
            name="height"
            value={formData.height}
            onChange={handleChange}
            placeholder="Enter height in cm"
            required
          />
        </div>

        <div className="form-group">
          <label htmlFor="weight">Weight (kg):</label>
          <input
            type="number"
            id="weight"
            name="weight"
            value={formData.weight}
            onChange={handleChange}
            placeholder="Enter weight in kg"
            required
          />
        </div>

        <div className="form-group">
          <label htmlFor="age">Age:</label>
          <input
            type="number"
            id="age"
            name="age"
            value={formData.age}
            onChange={handleChange}
            placeholder="Enter age"
            required
          />
        </div>

        <div className="form-group">
          <label htmlFor="eyeColor">Eye Color:</label>
          <select
            id="eyeColor"
            name="eyeColor"
            value={formData.eyeColor}
            onChange={handleChange}
            required
          >
            <option value="">Select eye color</option>
            <option value="brown">Brown</option>
            <option value="blue">Blue</option>
            <option value="green">Green</option>
            <option value="hazel">Hazel</option>
            <option value="other">Other</option>
          </select>
        </div>

        <button type="submit">Submit</button>
      </form>
    </div>
  )
}

export default App
```

Now let's add some basic styling. I'll create/modify the App.css file:

Filepath: src/App.css
Replace lines: 0-0
```css
#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
}

.form-container {
  max-width: 500px;
  margin: 0 auto;
  padding: 20px;
  background-color: #f5f5f5;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
}

h1 {
  text-align: center;
  color: #333;
  margin-bottom: 30px;
}

.form-group {
  margin-bottom: 20px;
}

label {
  display: block;
  margin-bottom: 5px;
  color: #333;
  font-weight: 500;
}

input, select {
  width: 100%;
  padding: 8px;
  border: 1px solid #ddd;
  border-radius: 4px;
  font-size: 16px;
}

input:focus, select:focus {
  outline: none;
  border-color: #646cff;
}

button {
  width: 100%;
  padding: 10px;
  background-color: #646cff;
  color: white;
  border: none;
  border-radius: 4px;
  font-size: 16px;
  cursor: pointer;
  transition: background-color 0.3s;
}

button:hover {
  background-color: #535bf2;
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

input[type=number] {
  -moz-appearance: textfield;
}
```



This creates a clean, user-friendly form with the following features:

1. Input fields for height (in cm), weight (in kg), and age
2. A dropdown select for eye color with common options
3. Form validation (all fields are required)
4. Clean, modern styling with hover effects
5. Responsive design that works on all screen sizes
6. Form data is managed in state and logged to console on submit

The form data is collected in the formData state object and can be accessed when the form is submitted. Currently, it just logs to the console, but you can modify the handleSubmit function to send the data wherever you need it.

**Generation complete!** Please review the code suggestions above.

============================================================

  response inside `.ts` file, notice that:
  - you CAN NOT see text highlighting
- you can see the three back ticks
- you can see the meta data

<img width="934" alt="Image" src="https://github.com/user-attachments/assets/fad0d212-010a-456a-a9b5-4e2a796a03dc" />

---

- Datetime: 2025-01-26 13:08:44

- Model: claude/claude-3-5-sonnet-20241022

- Selected files:
  - list.ts

> add a color property

I'll add a color property to the List class:

Filepath: list.ts
Replace lines: 7-13
```typescript
class List {
  private items: Item[];
  private color: string;

  constructor(color: string) {
    this.items = [];
    this.color = color;
  }

  getColor(): string {
    return this.color;
  }
```

The changes:
1. Added a private color property to the List class
2. Modified the constructor to require a color parameter
3. Added a getColor() method to access the color value

Now when creating a new List, you'll need to provide a color:
const list = new List("#FF0000"); // Red color in hex

============================================================

response inside `.rs` file, notice that:
- there is syntax highlighting
- the three back ticks are visible
- metadate at the top is visible. Datetime, model, selected files...


<img width="1188" alt="Image" src="https://github.com/user-attachments/assets/869cc7fb-bfc8-4d56-8daf-39bb3868933e" />

- Datetime: 2025-01-26 13:17:56

- Model: claude/claude-3-5-sonnet-20241022

- Selected files:
  - src/main.rs

> we need a function that takes in a string hopeAndDreams and then outputs "Idk sounds pretty difficult you should probably just stay home instead"

I'll help add this function to the Rust code. Let's add a new function that takes a String parameter and returns the discouraging message.

Here's the change:

Filepath: src/main.rs
Replace lines: 1-3
```rust
fn crush_hopes_and_dreams(hope_and_dreams: String) -> String {
    "Idk sounds pretty difficult you should probably just stay home instead".to_string()
}

fn main() {
    println!("Hello, world!");
    let dream = String::from("I want to climb Mount Everest");
    let response = crush_hopes_and_dreams(dream);
    println!("{}", response);
}
```

I've made the following changes:
1. Added a new function `crush_hopes_and_dreams` that takes a String parameter
2. The function returns a discouraging message as a String
3. Added an example usage in the main function to demonstrate how it works

Note that I followed Rust naming conventions by using snake_case for the function name rather than camelCase. The function will always return the same message regardless of the input, as per the requirement.

============================================================


`git diff:`
```
diff --git a/lua/avante/providers/claude.lua b/lua/avante/providers/claude.lua
index 469c93b..6ff5c82 100644
--- a/lua/avante/providers/claude.lua
+++ b/lua/avante/providers/claude.lua
@@ -77,6 +77,28 @@ M.parse_messages = function(opts)
   return messages
 end
 
+M.clean_response = function(text)
+  -- Remove metadata header more aggressively
+  text = text:gsub("^%-%-%-%s*\n[^>]*\n%-%-%-%s*\n", "")
+
+  -- Handle more backtick variations
+  text = text:gsub("```[%w%s]*\n", "")
+  text = text:gsub("\n```", "")
+
+  return text
+end
+
+-- M.clean_response = function(text)
+--   -- Remove metadata header
+--   text = text:gsub("^%-%-%-%s*\n.-\n%-%-%-%s*\n", "")
+--
+--   -- Remove backticks from code blocks
+--   text = text:gsub("```%w*\n", "")
+--   text = text:gsub("```%s*\n", "")
+--
+--   return text
+-- end
+
 M.parse_response = function(data_stream, event_state, opts)
   if event_state == nil then
     if data_stream:match('"content_block_delta"') then
@@ -85,10 +107,14 @@ M.parse_response = function(data_stream, event_state, opts)
       event_state = "message_stop"
     end
   end
+
   if event_state == "content_block_delta" then
     local ok, json = pcall(vim.json.decode, data_stream)
     if not ok then return end
-    opts.on_chunk(json.delta.text)
+    if json.delta and json.delta.text then
+      local cleaned_text = M.clean_response(json.delta.text)
+      opts.on_chunk(cleaned_text)
+    end
   elseif event_state == "message_stop" then
     opts.on_complete(nil)
     return
diff --git a/lua/avante/selection.lua b/lua/avante/selection.lua
index 37d5de7..ae1fb50 100644
--- a/lua/avante/selection.lua
+++ b/lua/avante/selection.lua
@@ -157,22 +157,33 @@ function Selection:create_editing_input()
 
     self.prompt_input:start_spinner()
 
+    ---@type AvanteChunkParser
     ---@type AvanteChunkParser
     local on_chunk = function(chunk)
       full_response = full_response .. chunk
       local response_lines_ = vim.split(full_response, "\n")
       local response_lines = {}
-      for i, line in ipairs(response_lines_) do
-        if string.match(line, "^```") and (i == 1 or i == #response_lines_) then goto continue end
-        if string.match(line, "^```$") then goto continue end
-        table.insert(response_lines, line)
-        ::continue::
+
+      -- Skip metadata header
+      local i = 1
+      while
+        i <= #response_lines_
+        and (
+          response_lines_[i]:match("^%-%-%-")
+          or response_lines_[i]:match("^%s*%-[%s]*Model:")
+          or response_lines_[i]:match("^%s*%-[%s]*Datetime:")
+          or response_lines_[i]:match("^%s*%-[%s]*Selected files:")
+        )
+      do
+        i = i + 1
       end
-      if #response_lines == 1 then
-        local first_line = response_lines[1]
-        local first_line_indentation = Utils.get_indentation(first_line)
-        need_prepend_indentation = first_line_indentation ~= original_first_line_indentation
+
+      -- Process remaining lines
+      for j = i, #response_lines_ do
+        local line = response_lines_[j]
+        if not (line:match("^```") or line:match("^```%w+$")) then table.insert(response_lines, line) end
       end
+
       if need_prepend_indentation then
         for i, line in ipairs(response_lines) do
           response_lines[i] = original_first_line_indentation .. line
```

**Comment by sudo-tee** [comment]
As of this morning it started to happen again even with the openai fix I posted above.

**Comment by currenthandle** [comment]
> As of this morning it started to happen again even with the openai fix I posted above.

@sudo-tee I write to you from the future where Avante works on my system once again (perfect or *almost* perfectly). We persevere 🫡

I can not suggest anyone follow in my footsteps here but this seemed to work for me
## Solution [TL:DR (Salt the Earth. Throw you computer in the ocean, get a new one and start over; NOT QUITE!)]

Update nvim
```
brew upgrade neovim
```

Nuke everything:
```
rm -rf ~/.local/share/nvim
rm -rf ~/.local/state/nvim
rm -rf ~/.cache/nvim
```

Open nvim and sync
```
nvim
:Lazy sync
```

Manually reinstall all all LSPs, linters, and formatters with `:Mason`.

Other than monkeying with a custom gotmp treesitter parser. That is pretty much all I did.

## Description of the issue
I was coding along happy as a clam me and Avante were really doing the thing. Then all the sudden a response came back malformed. Specifically, I was seeing `<SEARCH>` and `<REPLACE>` tags and code snippet replacements we not working. MAJOR BUMMER.

I created a blank file in my repo at my current working dir `touch test.ts`

I asked avante to create a really simple function in this file and it so so. Then I asked it to add a function. That's when it failed...

### And here is where it gets weird.
I copied that file to my home dir 
````
cp test.ts ~
cd ~
nvim test.ts
```

and then I opened that new `test.ts` file and tried to add a new function to it with Avante, it an worked, spooky 👻.

From here I started nuking everything.

## Everything seems to work now.
I am Avante-in pretty hard once again. Every once in a while I get a dreaded that is something like `E21 change edit because buffer is not 'modifiable'`. I'm sure I'm remembering the text wrong. But I am not yet sure if this is something I am doing wrong or if its just a bug that pops up now and again. I will continue to monitor it. But for now I can just close and re-open nvim. So things are pretty peachy.

Hope that helps.

**Comment by MrFox131** [comment]
> 1. No output.

I think may be related to #1251

**Comment by MrFox131** [comment]
> As of this morning it started to happen again even with the openai fix I posted above. 
> 
> 

I believe it's related to the issue mentioned above. Maybe u could have a look at it and create a pull request. I'm not really good at lua so I'm not gonna do it myself for now.

**Comment by synic** [comment]
> > As of this morning it started to happen again even with the openai fix I posted above. 
> > 
> > 
> 
> I believe it's related to the issue mentioned above. Maybe u could have a look at it and create a pull request. I'm not really good at lua so I'm not gonna do it myself for now.

If it really is the provider, then it might not be related. I'm using anthropic
